src/game/logic/object/object.cpp

```diff
@@ -2655,51 +2984,468 @@ void Object::Enter_Group(AIGroup *group)
+            m_threatSighting->m_frame,
+            m_threatSighting->m_playerIndex);
+        m_threatSighting->Reset();
+    }
 }
 
-void Object::Topple(const Coord3D dir, float speed, int options) {}
+void Object::Look()
+{
+    if (m_friendlyLookSighting->Is_Invalid()) {
+        if (Get_Controlling_Player() != nullptr && !Is_Destroyed() && !Is_Effectively_Dead()) {
+            ContainModuleInterface *contain;
+
+            if (Get_Contained_By() != nullptr) {
+                contain = Get_Contained_By()->Get_Contain();
+            } else {
+                contain = nullptr;
+            }
+
+            if (contain == nullptr || contain->Is_Garrisonable()) {
+                float radius = Get_Shroud_Clearing_Range();
+
+                if (radius > 0.0f) {
+                    unsigned short index = 0;
+                    unsigned short index2;
+
+                    if (Is_KindOf(KINDOF_REVEAL_TO_ALL)) {
+                        index2 = -1;
+                    } else {
+                        for (int i = g_thePlayerList->Get_Player_Count() - 1; i >= 0; i--) {
+                            Player *player = g_thePlayerList->Get_Nth_Player(i);
+
+                            if (Get_Controlling_Player()->Get_Relationship(player->Get_Default_Team()) == ALLIES) {
+                                index |= player->Get_Player_Mask();
+                            }
+                        }
+
+                        index2 = m_spiedOnPlayers | index;
+                    }
+
+                    Coord3D pos = *Get_Position();
+                    g_thePartitionManager->Do_Shroud_Reveal(pos.x, pos.y, radius, index2);
+                    m_friendlyLookSighting->m_where = pos;
+                    m_friendlyLookSighting->m_playerIndex = index2;
+                    m_friendlyLookSighting->m_radius = Get_Shroud_Clearing_Range();
+                }
+
+                float range = Get_Template()->Get_Shroud_Reveal_To_All_Range();
+
+                if (range > 0.0f && !Get_Status(OBJECT_STATUS_UNDER_CONSTRUCTION)
+                    && (!Get_Status(OBJECT_STATUS_STEALTHED) || Get_Status(OBJECT_STATUS_DETECTED)
+                        || Get_Status(OBJECT_STATUS_DISGUISED))) {
+                    Coord3D pos = *Get_Position();
+                    unsigned short players =
+                        g_thePlayerList->Get_Players_With_Relationship(Get_Controlling_Player()->Get_Player_Mask(),
+                            PLAYER_RELATIONSHIP_FLAGS_ENEMIES | PLAYER_RELATIONSHIP_FLAGS_NEUTRALS);
+                    g_thePartitionManager->Do_Shroud_Reveal(pos.x, pos.y, range, players);
+                    m_allLookSighting->m_where = pos;
+                    m_allLookSighting->m_playerIndex = players;
+                    m_allLookSighting->m_radius = range;
+                }
+            }
+        }
+    } else {
+        captainslog_dbgassert(false, "An Object is looking, but hasn't unlooked the last one.");
```

Comment:
Can add object name into this message. Otherwise this assert does not tell much.